docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,639 @@
+   */
+  onUpdate(callback: (payload: HMRPayload) => void): void
+}
+```
+
+This interface defines how HMR communication is established. Vite exports `ServerHMRConnector` from the main entry point to support HMR during Vite SSR. The `isReady` and `send` methods are usually called when the custom event is triggered (like, `import.meta.hot.send("my-event")`).
+
+`onUpdate` is called only once when the new runtime is initiated. It passed down a method that should be called when connection triggers the HMR event. The implementation depends on the type of connection (as an example, it can be `WebSocket`/`EventEmitter`/`MessageChannel`), but it usually looks something like this:
+
+```js
+function onUpdate(callback) {
+  this.connection.on('hmr', (event) => callback(event.data))
+}
+```
+
+The callback is queued and it will wait for the current update to be resolved before processing the next update. Unlike the browser implementation, HMR updates in Vite Runtime wait until all listeners (like, `vite:beforeUpdate`/`vite:beforeFullReload`) are finished before updating the modules.
+
+## Environments during build
+
+Plugin hooks also receive the environment name during build. This replaces the `ssr` boolean we have been passing them so far.
+
+The Vite CLI would aslo be updated from `vite build --ssr` to `vite build --environment=ssr`. Applications can then call build for each environment (for example `vite build --environment=workerd`).
+
+In a future stage, or as part of this proposal, we could also review or stance on vite build being a simple wrapper around rollup. Instead, now that we have a proper environment concept, vite build could create a `ViteBuilder` that has knowledge of all the configured environments and build them all with a single call. This has been requested many times by framework authors that use the "Framework as a plugin" scheme. Right now they end up triggering the SSR build when the `buildEnd` hook for the browser build is called.
+
+In its simpler form, the vite builder could call each build in series as defined by `config.environments` order (or by a new `config.build.environments` order). This should cover most current use cases out of the box, given that most frameworks build the client first and then SSR (as they use the client manifest).
+
+It is an interesting design space to explore because it could make build and dev work in a more uniform way. For example, Vite Builder could also only load the config file once and do the overrides for each environment in the exact same way as it is done during dev. And there could also be a single plugin pipeline, if we would like plugins to share state directly between the environments (as it happens during dev already).
+
+## Backward Compatibility
+
+The current Vite server API will be deprecated but keep working during the next major.
+
+- `server.transformRequest(url)` <br> -> `server.environment('browser').transformRequest(url)`
```

Comment:
I was afraid it will end up being too wide, but it looks good. Pushed. Thanks!

<img width="756" alt="image" src="https://github.com/vitejs/vite/assets/583075/12f01786-bd07-4966-b22a-f75318cf2491">